README.md
# Bedrock OAuth2 Dynamic Client Registration _(@digitalbazaar/oauth2-client-registration-handler)_

[![Node.js CI](https://github.com/digitalbazaar/oauth2-client-registration-handler/workflows/Node.js%20CI/badge.svg)](https://github.com/digitalbazaar/oauth2-client-registration-handler/actions?query=workflow%3A%22Node.js+CI%22)

> OAuth2 Dynamic Registration handler for custom authorization servers.

## Table of Contents

- [Background](#background)
- [Security](#security)
- [Install](#install)
- [Usage](#usage)
- [Contribute](#contribute)
- [Commercial Support](#commercial-support)
- [License](#license)

## Background

Constraints/Limitations:

* Only supporting `client_credentials` grant type for now.
* Initial registration authentication method is bearer token.
* Does not support the signed Software Statement mechanism.

Relevant specifications:

* [OAuth 2.0 Dynamic Client Registration Protocol](https://tools.ietf.org/html/rfc7591)
* [OAuth 2.0 Dynamic Client Registration Management Protocol](https://tools.ietf.org/html/rfc7592)

## Security

TBD

## Install

- Node.js >= 16 is required.

To install locally (for development):

```
git clone https://github.com/digitalbazaar/oauth2-client-registration-handler.git
cd oauth2-client-registration-handler
npm install
```

## Usage

This library exports a route handler to perform OAuth2 Dynamic Client
Registration that can be added to an existing Bedrock or Express.js application.

```js
import {clientRegistrationHandler} from '@digitalbazaar/oauth2-client-registration-handler';

app.post('/oauth2/register',
  clientRegistrationHandler({
    baseUrl: 'https://as.example.com',
    authentication: {
      strategy: 'bearer',
      validateInitialAccessToken: async ({token}) => {/* custom token validation logic */}
    },
    register: async ({registration, credentials: {initialAccessToken}}) => {
      // custom registration callback (saves client to database etc)
    },
    defaults: {
      clientSecretExpiresAt: 0, // never expires
      grantTypes: ['client_credentials'],
      tokenEndpointAuthMethod: 'client_secret_post'
    },
    logger: console
  }))
);
```

## Contribute

See [the contribute file](https://github.com/digitalbazaar/bedrock/blob/master/CONTRIBUTING.md)!

PRs accepted.

If editing the Readme, please conform to the
[standard-readme](https://github.com/RichardLitt/standard-readme) specification.

## Commercial Support

Commercial support for this library is available upon request from
Digital Bazaar: support@digitalbazaar.com

## License

[New BSD License (3-clause)](LICENSE) © Digital Bazaar


CHANGELOG.md
# oauth2-client-registration-handler ChangeLog

## 2.0.0 - 2023-02-06

### Changed
- **BREAKING**: Convert to module (ESM).
- **BREAKING**: Drop support for node <= 14.
- Update dev deps to latest.
- Update `bnid` to v3.0 (Converted to ESM module).
- Use `c8@7.12.0` for coverage.

## 1.0.0 - 2020-10-09

- Initial release. See git history for changes.


test/test.js
/*
 * Copyright (c) 2019-2020 Digital Bazaar, Inc. All rights reserved.
 */
const bedrock = require('bedrock');
require('bedrock-https-agent');
require('bedrock-kms-http');
require('bedrock-security-context');
require('bedrock-module-template-http');

// this is responsible for providing the `ssm-v1` key store
require('bedrock-ssm-mongodb');

require('bedrock-test');
bedrock.start();


test/test.config.js
/*
 * Copyright (c) 2019-2020 Digital Bazaar, Inc. All rights reserved.
 */
'use strict';

const {config} = require('bedrock');
const path = require('path');

// MongoDB
config.mongodb.name = 'bedrock_module_template_http_test';
config.mongodb.dropCollections.onInit = true;
config.mongodb.dropCollections.collections = [];

// do not require an authentication session for tests
config['kms-http'].requireAuthentication = false;

config.mocha.tests.push(path.join(__dirname, 'mocha'));

config.kms.allowedHost = config.server.host;

// allow self-signed certs in test framework
config['https-agent'].rejectUnauthorized = false;


tests/index.spec.js
/*!
 * Copyright (c) 2020-2023 Digital Bazaar, Inc. All rights reserved.
 */
import {_parseRegistration, clientRegistrationHandler} from '../lib/index.js';
import chai from 'chai';
import chaiHttp from 'chai-http';
import express from 'express';

chai.use(chaiHttp);
chai.should();
const {expect} = chai;

const bodyParserJson = express.json();

const registerUrl = '/oauth2/register';
const VALID_TOKEN = 'abcd1234';
const INVALID_TOKEN = 'xyz567';

describe('handleClientRegistration', () => {
  const app = express();
  app.post(
    registerUrl,
    bodyParserJson,
    clientRegistrationHandler({
      baseUrl: 'https://as.example.com',
      authentication: {
        strategy: 'bearer',
        validateInitialAccessToken: async ({token}) => {
          if(token !== VALID_TOKEN) {
            throw new Error('Token not found.');
          }
        }
      },
      // eslint-disable-next-line no-unused-vars
      register: async ({registration, credentials: {initialAccessToken}}) => {
        // save to database, then return the saved client metadata
        return registration;
      }
    })
  );

  let requester;

  before(async () => {
    requester = chai.request(app).keepOpen();
  });

  after(async () => {
    requester.close();
  });

  it('should error if called un-authenticated', async () => {
    const res = await requester.post(registerUrl).send({});
    expect(res).to.have.status(403);
    expect(res).to.be.json;
    expect(res.body.error).to.equal('access_denied');
    expect(res.body.error_description)
      .to.equal('Authentication Code required.');
  });

  it('should error if called with a non-bearer auth scheme', async () => {
    const res = await requester.post(registerUrl)
      .set('Authorization', 'Basic abcd123')
      .send({});
    expect(res).to.have.status(400);
    expect(res).to.be.json;
    expect(res.body.error).to.equal('invalid_request');
    expect(res.body.error_description)
      .to.equal('Invalid authorization scheme.');
  });

  it('should error if given an invalid token', async () => {
    const res = await requester.post(registerUrl)
      .set('Authorization', `Bearer ${INVALID_TOKEN}`)
      .send({});
    expect(res).to.have.status(403);
    expect(res.body.error_description).to
      .equal('Invalid authentication code: Token not found.');
  });

  it('should return a valid registration with correct auth token', async () => {
    const res = await requester.post(registerUrl)
      .set('Authorization', `Bearer ${VALID_TOKEN}`)
      .send({
        client_name: 'Example client.'
      });
    expect(res).to.have.status(201);
    expect(res).to.be.json;
    expect(res.body.grant_types).to.eql(['client_credentials']);
    expect(res.body.client_id).to.be.a('string');
    expect(res.body.client_secret).to.be.a('string');
    expect(res.body.client_id_issued_at).to.be.a('number');
    expect(res.body.client_secret_expires_at).to.equal(0);
    expect(res.body.token_endpoint_auth_method).to.equal('client_secret_post');
    expect(res.body.client_name).to.equal('Example client.');
  });
});

describe('_parseRegistration', () => {
  it('should throw if encounters a forbidden field', async () => {
    let error;
    try {
      _parseRegistration({
        registrationBody: {client_secret: 'some client secret'}
      });
    } catch(e) {
      error = e;
    }
    expect(error.message).to.equal('invalid_request');
    expect(error.error_description).to
      .equal('Registration MUST NOT include the "client_secret" field.');
  });
});


lib/main.js
/*!
 * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
 */
import * as authorization from 'auth-header';
import ClientMetadata from './ClientMetadata';
import {generateId} from 'bnid';
import {InvalidRequest, AccessDenied} from '@interop/oauth2-errors';
import {snakeCase} from 'snake-case';

const FORBIDDEN_FIELDS = new Set([
  'registration_access_token',
  'registration_client_uri',
  'client_secret',
  'client_secret_expires_at',
  'client_id_issued_at',
]);

const REGISTRATION_DEFAULTS = {
  grant_types: ['client_credentials'],
  token_endpoint_auth_method: 'client_secret_post'
};

/**
 * Returns an Express.js/Bedrock middleware function that handles OAuth 2.0
 * Dynamic Client Registration.
 *
 * @see https://tools.ietf.org/html/rfc7591
 *
 * @param {object} options - Options hashmap.
 * @param {object} options.authentication - Authentication strategy config,
 *   including the endpoint authn strategy, and the validate initial token
 *   callback. For example,
 *   `{strategy: 'bearer', validateInitialAccessToken}`.
 * @param {Function} options.register - Registration callback with the signature
 *   `({registration, credentials: {initialAccessToken}})`, where the
 *   client registration is typically persisted in a database.
 * @param {object} [options.defaults={}] - Hashmap of default registration
 *   values.
 * @param {boolean} [options.allowClientProvidedId=false] - Whether to allow
 *   clients to provide their own `client_id` values in the registration.
 * @param {object} options.logger - Logger.
 *
 * @returns {Function} Returns middleware request handler.
 */
function clientRegistrationHandler({
  authentication, register, defaults = {}, allowClientProvidedId = false,
  logger
}) {
  return async (req, res) => {
    try {
      const initialAccessToken = await _authenticate({
        req, authentication, logger});

      const {body: registrationBody} = req;

      const client = _parseRegistration({registrationBody, defaults});

      const result = client.validateNew();
      if(!result.valid) {
        throw result.error;
      }

      // client can be optionally overriden by the register() callback
      await _ensureClientId({client, allowClientProvidedId});
      await _generateClientSecret({client, defaults});

      // Epoch time
      client.meta.client_id_issued_at = Math.floor(Date.now() / 1000);

      let registrationResponse;
      try {
        registrationResponse = await register({
          registration: client.meta, credentials: {initialAccessToken}
        });
      } catch(e) {
        throw new InvalidRequest({
          description: `Error in register() callback: "${e.message}".`
        });
      }

      return _respond({registrationResponse, res});
    } catch(error) {
      _error({error, logger, res});
    }
  };
}

function _error({error, logger, res}) {
  if(logger) {
    logger.error('Dynamic client registration error.', {error});
  }
  const {
    error: errorId, error_description: description, error_uri: uri
  } = error;

  const statusCode = error.statusCode || 400;
  const oauth2ErrorResponse = {
    error: errorId || 'invalid_request',
    error_description: description || error.message,
    error_uri: uri
  };
  res.set({
    'Cache-Control': 'no-store',
    Pragma: 'no-cache'
  });
  return res.status(statusCode).json(oauth2ErrorResponse);
}

function _respond({registrationResponse, res}) {
  res.set({
    'Cache-Control': 'no-store',
    Pragma: 'no-cache'
  });

  res.status(201).json(registrationResponse);
}

function _parseRegistration({registrationBody, defaults = {}}) {
  if(!registrationBody) {
    throw new InvalidRequest({
      description: 'Missing registration request body.'
    });
  }

  for(const field in registrationBody) {
    if(FORBIDDEN_FIELDS.has(field)) {
      throw new InvalidRequest({
        description: `Registration MUST NOT include the "${field}" field.`
      });
    }
  }

  const registrationDefaults = {...REGISTRATION_DEFAULTS};
  for(const [key, value] of Object.entries(defaults)) {
    registrationDefaults[snakeCase(key)] = value;
  }

  return new ClientMetadata({...registrationDefaults, ...registrationBody});
}

async function _authenticate({req, res, authentication}) {
  let token;

  if(authentication.strategy === 'bearer') {
    token = await _extractBearerToken({req, res});
    try {
      await authentication.validateInitialAccessToken({token});
    } catch(e) {
      throw new AccessDenied({
        description: `Invalid authentication code: ${e.message}`
      });
    }

  } else {
    throw new InvalidRequest({
      description: 'Only the Bearer auth scheme is currently supported.'});
  }
  return token;
}

async function _extractBearerToken({req}) {
  const header = req.get('Authorization');
  if(!header) {
    throw new AccessDenied({description: 'Authentication Code required.'});
  }

  const {scheme, token} = authorization.parse(header);

  if(!(scheme === 'Bearer' && token)) {
    throw new InvalidRequest({description: 'Invalid authorization scheme.'});
  }

  return token;
}

async function _ensureClientId({client, allowClientProvidedId}) {
  if(client.meta.client_id && allowClientProvidedId) {
    return;
  }

  client.meta.client_id = await generateId();
}

async function _generateClientSecret({client, defaults}) {
  if(client.requiresClientSecret()) {
    client.meta.client_secret = await generateId();
  }
  if(client.meta.client_secret) {
    client.meta.client_secret_expires_at = defaults.clientSecretExpiresAt || 0;
  }
}

export {
  FORBIDDEN_FIELDS,
  REGISTRATION_DEFAULTS,
  clientRegistrationHandler,
  ClientMetadata,
  _authenticate,
  _ensureClientId,
  _extractBearerToken,
  _generateClientSecret,
  _parseRegistration,
  _respond
};


lib/ClientMetadata.js
/*!
 * Copyright (c) 2020-2023 Digital Bazaar, Inc. All rights reserved.
 */
import {InvalidRequest} from '@interop/oauth2-errors';

class ClientMetadata {
  constructor(meta = {}) {
    this.meta = meta;
  }

  validateNew() {
    try {
      if(!this.meta.redirect_uris && this.redirectUriRequired()) {
        throw new InvalidRequest({
          description: 'Missing redirect_uris parameter.'
        });
      }
    } catch(error) {
      return {valid: false, error};
    }

    return {valid: true};
  }

  requiresClientSecret() {
    return !this.implicitFlow();
  }

  implicitFlow() {
    const responseTypes = this.meta.response_types;

    return !!(responseTypes &&
      responseTypes.length === 1 &&
      responseTypes[0] === 'id_token token');
  }

  redirectUriRequired() {
    // client_credentials grant type does not need redirect_uris
    return false;
  }
}

export default ClientMetadata;


lib/index.js
/*!
 * Copyright (c) 2020-2023 Digital Bazaar, Inc. All rights reserved.
 */
import * as authorization from 'auth-header';
import {AccessDenied, InvalidRequest} from '@interop/oauth2-errors';
import ClientMetadata from './ClientMetadata.js';
import {generateId} from 'bnid';
import {snakeCase} from 'snake-case';

const FORBIDDEN_FIELDS = new Set([
  'registration_access_token',
  'registration_client_uri',
  'client_secret',
  'client_secret_expires_at',
  'client_id_issued_at',
]);

const REGISTRATION_DEFAULTS = {
  grant_types: ['client_credentials'],
  token_endpoint_auth_method: 'client_secret_post'
};

/**
 * Returns an Express.js/Bedrock middleware function that handles OAuth 2.0
 * Dynamic Client Registration.
 *
 * @see https://tools.ietf.org/html/rfc7591
 *
 * @param {object} options - Options hashmap.
 * @param {object} options.authentication - Authentication strategy config,
 *   including the endpoint authn strategy, and the validate initial token
 *   callback. For example,
 *   `{strategy: 'bearer', validateInitialAccessToken}`.
 * @param {Function} options.register - Registration callback with the signature
 *   `({registration, credentials: {initialAccessToken}})`, where the
 *   client registration is typically persisted in a database.
 * @param {object} [options.defaults={}] - Hashmap of default registration
 *   values.
 * @param {boolean} [options.allowClientProvidedId=false] - Whether to allow
 *   clients to provide their own `client_id` values in the registration.
 * @param {object} options.logger - Logger.
 *
 * @returns {Function} Returns middleware request handler.
 */
function clientRegistrationHandler({
  authentication, register, defaults = {}, allowClientProvidedId = false,
  logger
}) {
  return async (req, res) => {
    try {
      const initialAccessToken = await _authenticate({
        req, authentication, logger});

      const {body: registrationBody} = req;

      const client = _parseRegistration({registrationBody, defaults});

      const result = client.validateNew();
      if(!result.valid) {
        throw result.error;
      }

      // client can be optionally overriden by the register() callback
      await _ensureClientId({client, allowClientProvidedId});
      await _generateClientSecret({client, defaults});

      // Epoch time
      client.meta.client_id_issued_at = Math.floor(Date.now() / 1000);

      let registrationResponse;
      try {
        registrationResponse = await register({
          registration: client.meta, credentials: {initialAccessToken}
        });
      } catch(e) {
        throw new InvalidRequest({
          description: `Error in register() callback: "${e.message}".`
        });
      }

      return _respond({registrationResponse, res});
    } catch(error) {
      _error({error, logger, res});
    }
  };
}

function _error({error, logger, res}) {
  if(logger) {
    logger.error('Dynamic client registration error.', {error});
  }
  const {
    error: errorId, error_description: description, error_uri: uri
  } = error;

  const statusCode = error.statusCode || 400;
  const oauth2ErrorResponse = {
    error: errorId || 'invalid_request',
    error_description: description || error.message,
    error_uri: uri
  };
  res.set({
    'Cache-Control': 'no-store',
    Pragma: 'no-cache'
  });
  return res.status(statusCode).json(oauth2ErrorResponse);
}

function _respond({registrationResponse, res}) {
  res.set({
    'Cache-Control': 'no-store',
    Pragma: 'no-cache'
  });

  res.status(201).json(registrationResponse);
}

function _parseRegistration({registrationBody, defaults = {}}) {
  if(!registrationBody) {
    throw new InvalidRequest({
      description: 'Missing registration request body.'
    });
  }

  for(const field in registrationBody) {
    if(FORBIDDEN_FIELDS.has(field)) {
      throw new InvalidRequest({
        description: `Registration MUST NOT include the "${field}" field.`
      });
    }
  }

  const registrationDefaults = {...REGISTRATION_DEFAULTS};
  for(const [key, value] of Object.entries(defaults)) {
    registrationDefaults[snakeCase(key)] = value;
  }

  return new ClientMetadata({...registrationDefaults, ...registrationBody});
}

async function _authenticate({req, res, authentication}) {
  let token;

  if(authentication.strategy === 'bearer') {
    token = await _extractBearerToken({req, res});
    try {
      await authentication.validateInitialAccessToken({token});
    } catch(e) {
      throw new AccessDenied({
        description: `Invalid authentication code: ${e.message}`
      });
    }

  } else {
    throw new InvalidRequest({
      description: 'Only the Bearer auth scheme is currently supported.'});
  }
  return token;
}

async function _extractBearerToken({req}) {
  const header = req.get('Authorization');
  if(!header) {
    throw new AccessDenied({description: 'Authentication Code required.'});
  }

  const {scheme, token} = authorization.parse(header);

  if(!(scheme === 'Bearer' && token)) {
    throw new InvalidRequest({description: 'Invalid authorization scheme.'});
  }

  return token;
}

async function _ensureClientId({client, allowClientProvidedId}) {
  if(client.meta.client_id && allowClientProvidedId) {
    return;
  }

  client.meta.client_id = await generateId();
}

async function _generateClientSecret({client, defaults}) {
  if(client.requiresClientSecret()) {
    client.meta.client_secret = await generateId();
  }
  if(client.meta.client_secret) {
    client.meta.client_secret_expires_at = defaults.clientSecretExpiresAt || 0;
  }
}

export {
  FORBIDDEN_FIELDS,
  REGISTRATION_DEFAULTS,
  clientRegistrationHandler,
  ClientMetadata,
  _authenticate,
  _ensureClientId,
  _extractBearerToken,
  _generateClientSecret,
  _parseRegistration,
  _respond
};
